k2/csrc/fsa.h

```diff
@@ -77,13 +78,40 @@ struct Fsa {
   // contains the first arc-index leaving this state (index into `arcs`).
   // The next element of this array gives the end of that range.
   // Note: the final-state is numbered last, and implicitly has no
-  // arcs leaving it.
+  // arcs leaving it. For non-empty FSA, we put a duplicate of the final state
+  // at the end of `arc_indexes` to avoid boundary check for some FSA
+  // operations. Caution: users should never call `arc_indexes.size()` to get
+  // the number of states, they should call `NumStates()` to get the number.
   std::vector<int32_t> arc_indexes;
 
   // Note: an index into the `arcs` array is called an arc-index.
   std::vector<Arc> arcs;
 
-  StateId NumStates() const { return static_cast<StateId>(arc_indexes.size()); }
+  Fsa() = default;
+  // just for creating testing FSA examples for now.
+  Fsa(std::vector<Arc> fsa_arcs, int32_t final_state)
+      : arcs(std::move(fsa_arcs)) {
+    if (arcs.empty()) return;
+
+    int32_t curr_state = -1;
+    int32_t index = 0;
+    for (const auto &arc : arcs) {
+      while (curr_state < arc.src_state) {
+        arc_indexes.push_back(index);
+        ++curr_state;
+      }
+      ++index;
+    }
+    // noted that here we push two `final_state` at the end, the last element is
+    // just to avoid boundary check for some FSA operations.
+    for (; curr_state <= final_state; ++curr_state)
+      arc_indexes.push_back(index);
```

Comment:
braces {} ?

Reply:
so one-line for-loop should be with `{}` but one-line if is not? It's strange.